src/dataflow/src/render/delta_join.rs

```diff
@@ -341,146 +382,224 @@ where
+        mfp: &mut MapFilterProject,
+    ) -> Self {
+        // First, determine which columns should be compare due to `equivalences`.
+        let mut ready_equivalences = Vec::new();
+        for equivalence in equivalences.iter_mut() {
+            if let Some(pos) = equivalence
+                .iter()
+                .position(|e| e.support().into_iter().all(|c| columns.contains_key(&c)))
+            {
+                let mut should_equate = Vec::new();
+                let mut cursor = pos + 1;
+                while cursor < equivalence.len() {
+                    if equivalence[cursor]
+                        .support()
+                        .into_iter()
+                        .all(|c| columns.contains_key(&c))
+                    {
+                        // Remove expression and equate with the first bound expression.
+                        should_equate.push(equivalence.remove(cursor));
+                    } else {
+                        cursor += 1;
+                    }
                 }
-            });
-        }
-    }
-
-    // Apply a filter if either list of constraints is non-empty.
-    if ready_to_go.is_empty() && ready_equivalences.is_empty() {
-        (updates, None)
-    } else {
-        let (ok_collection, err_collection) = updates.filter_fallible(move |input_row| {
-            let temp_storage = repr::RowArena::new();
-            let datums = input_row.unpack();
-            for p in &ready_to_go {
-                if p.eval(&datums, &temp_storage)? != Datum::True {
-                    return Ok(false);
+                if !should_equate.is_empty() {
+                    should_equate.push(equivalence[pos].clone());
+                    ready_equivalences.push(should_equate);
                 }
             }
-            for exprs in &ready_equivalences {
-                // Each list of expressions should be equal to the same value.
-                let val = exprs[0].eval(&datums, &temp_storage)?;
-                for expr in exprs[1..].iter() {
-                    if expr.eval(&datums, &temp_storage)? != val {
-                        return Ok(false);
+        }
+        equivalences.retain(|e| e.len() > 1);
+
+        // Update ready_equivalences to reference correct column locations.
+        for exprs in ready_equivalences.iter_mut() {
+            for expr in exprs.iter_mut() {
+                expr.visit_mut(&mut |e| {
+                    if let ScalarExpr::Column(c) = e {
+                        *c = columns[c];
                     }
-                }
+                });
+            }
+        }
+
+        // Next, partition `mfp` into `before` and `after`, the former of which can be
+        // applied now.
+        let (mut before, after) = std::mem::replace(mfp, MapFilterProject::new(mfp.input_arity))
+            .partition(columns, columns.len());
+
+        // While it is top of mind, add any new columns to `columns` and increase `input_arity`.
```

Comment:
It's not clear to me where the second part of this comment ( "increase input_arity" ) happens

Reply:
Good point! It used to be relevant, and ceased to be so. `input_arity` is always `columns.len()` in this method.